<a href="https://colab.research.google.com/github/szeshiang/Deep-Learning/blob/main/3_Mar_24_CNN_Ex_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# load the data
from tensorflow.keras.datasets import mnist
(X_train,y_train),(X_test,y_test)=mnist.load_data()
print('Data loaded')

11490434/11490434 [==============================] - 0s 0us/step
Data loaded


In [2]:
X_train.shape

(60000, 28, 28)

In [3]:
# reshape data to reflect the 3rd dimension (colour). In this dataset, the colour is B&W.
X_train=X_train.reshape(60000,28,28,1)
X_test=X_test.reshape(10000,28,28,1)

In [4]:
X_train.shape

(60000, 28, 28, 1)

In [5]:
# convert y into one-hot labels
num_classes=10
from tensorflow.keras.utils import to_categorical
y_train=to_categorical(y_train,num_classes)
y_test=to_categorical(y_test,num_classes)

In [6]:
y_train[:5]

array([[0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.]], dtype=float32)

In [7]:
# create CNN architecture
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense,Conv2D,MaxPooling2D,Flatten,Dropout
print('Imported')

Imported


In [8]:
# create a CNN with 2-layers
model=Sequential()
# first CNN layer
model.add(Conv2D(64,kernel_size=(3,3),activation='relu', input_shape=(28,28,1)))  # no of filters (64, 32, 28, etc), filter window (kernel size)
model.add(MaxPooling2D(pool_size=(2,2)))

# second CNN layer
model.add(Conv2D(32,kernel_size=(3,3),activation='relu'))
model.add(MaxPooling2D(pool_size=(2,2)))

# Flatten (for ANN)
model.add(Flatten())

# create ANN layer
model.add(Dense(1000, activation='relu'))

# output layer
model.add(Dense(10,activation='softmax'))

In [9]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 64)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 32)        18464     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 32)          0         
 g2D)                                                            
                                                                 
 flatten (Flatten)           (None, 800)               0         
                                                                 
 dense (Dense)               (None, 1000)              8

**1st CNN shape** = input shape - (minus) (kernel-1)
                    28-(3-1) = **26** <-- shape of first CNN layer

**max pooling (2x2)** --> 26 reduce to 13

**param of 1st layer** = (shape of filter+bias)*num of filters
                         ((3x3)+1)x64 = 640

----------------------------------------------------------
**2nd CNN shape** = 13-(3-1) = 11

**max pooling (2x2)** --> 11 reduce to 5

**param of 2nd layer** = (((3x3)x64 of previous output)+1)*32 = 18464

In [10]:
# compile the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
print('compiled')

compiled


In [11]:
# run
history=model.fit(X_train,y_train,epochs=3,batch_size=128,
                  validation_data=(X_test,y_test))

Epoch 1/3
469/469 [==============================] - 9s 7ms/step - loss: 0.7117 - accuracy: 0.9305 - val_loss: 0.0715 - val_accuracy: 0.9781
Epoch 2/3
469/469 [==============================] - 3s 5ms/step - loss: 0.0584 - accuracy: 0.9822 - val_loss: 0.0801 - val_accuracy: 0.9776
Epoch 3/3
469/469 [==============================] - 3s 6ms/step - loss: 0.0389 - accuracy: 0.9877 - val_loss: 0.0645 - val_accuracy: 0.9799


In [12]:
# evaluate the model
model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 3ms/step - loss: 0.0645 - accuracy: 0.9799


[0.06449466198682785, 0.9799000024795532]

In [13]:
model.evaluate(X_train, y_train)

1875/1875 [==============================] - 5s 3ms/step - loss: 0.0289 - accuracy: 0.9903


[0.02893577702343464, 0.9902833104133606]